toolkit/tools/graphpkgfetcher/graphpkgfetcher.go

```diff
@@ -209,21 +292,149 @@ func resolveGraphNodes(dependencyGraph *pkggraph.PkgGraph, inputSummaryFile, out
+//   - dependencyGraphDeltaCopy: A copy of the graph we will use to try to optimize the build nodes. This graph should be
+//     optimized to only contain the nodes we need to build.
+//   - cloner: The cloner to use to download the RPMs
+func downloadAllAvailableDeltaRPMs(realDependencyGraph, dependencyGraphDeltaCopy *pkggraph.PkgGraph, cloner *rpmrepocloner.RpmRepoCloner) (err error) {
+	timestamp.StartEvent("downloading delta nodes", nil)
+	defer timestamp.StopEvent(nil)
+
+	// First scan the copy of the graph we tried to optimize for all the SRPMs we need to build. We will use this list to
+	// match against all the nodes in the full graph.
+	srpmPaths := make(map[string]bool)
+	for _, n := range dependencyGraphDeltaCopy.AllBuildNodes() {
+		srpmPaths[n.SrpmPath] = true
+	}
+
+	// For each build node, try to update it to a delta node with a downloaded RPM backing it.
+	logger.Log.Debugf("Resolving build nodes")
+	buildNodes := realDependencyGraph.AllBuildNodes()
+	for i, n := range buildNodes {
+		progress := (i * 100) / len(buildNodes)
+		// If this node isn't part of the optimized graph, skip it.
+		if _, ok := srpmPaths[n.SrpmPath]; !ok {
+			logger.Log.Debugf("Skipping non-optimized delta build node %s", n)
+			continue
+		}
+
+		logger.Log.Debugf("Resolving build node %s", n)
+		if n.State == pkggraph.StateBuild {
+			err := downloadSingleDeltaRPM(realDependencyGraph, n, cloner, *outDir, progress)
+			if err != nil {
+				return fmt.Errorf("failed to download delta RPM for build node %s: %w", n, err)
+			}
 		}
 	}
 
 	return
 }
 
+// downloadSingleDeltaRPM attempts to download a single delta RPM for a build node. If the delta RPM is available
+// it will be downloaded and the build node will be updated to point to the new RPM. The associated run node will
+// also be updated to point to the new RPM since the scheduler uses the run node to find the RPM to install. If a
+// delta RPM is not available, the build node will be left alone an no error will be returned.
+//   - realDependencyGraph: The graph to update
+//   - buildNode: The build node to update. This node should be from the real graph as we will be updating it directly.
+//     to find the actual build node in the graph.
+//   - cloner: The cloner to use to download the RPMs
+//   - deltaRpmDir: The directory to download the RPMs into
+func downloadSingleDeltaRPM(realDependencyGraph *pkggraph.PkgGraph, buildNode *pkggraph.PkgNode, cloner *rpmrepocloner.RpmRepoCloner, deltaRpmDir string, progress int) (err error) {
+	const downloadDependencies = false
+	var lookup *pkggraph.LookupNode
+
+	if buildNode.Type != pkggraph.TypeBuild {
+		err = fmt.Errorf("node '%s' is not a build node, can't download delta RPM", buildNode)
+		return err
+	}
+
+	// Replace all '/' with '_' in the package name to get a valid timestamp name
+	// e.g. "bin/ls" -> "bin_ls"
+	tsName := strings.ReplaceAll(buildNode.VersionedPkg.Name, "/", "_")
```

Comment:
Seeing how `StartEvent` accepts strings with spaces in them, I feel that either `StartEvent` should accept all possible strings or make the necessary conversions itself. We can leave this code here for now, but I'd log a bug to address this issue in a separate PR.